File: QuantumVerse_Organized/03_Database/schemas/learning.py
import uuid
from datetime import datetime
from sqlalchemy import String, Integer, Boolean, ForeignKey, Text, JSON, DateTime, Enum
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database.connection import Base
import enum

class ProgressStatus(str, enum.Enum):
    not_started = "not_started"
    in_progress = "in_progress"
    completed = "completed"

class LearningModule(Base):
    __tablename__ = "learning_modules"
    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str] = mapped_column(Text, default="")
    level: Mapped[int] = mapped_column(Integer, nullable=False)  # 1-4
    order_index: Mapped[int] = mapped_column(Integer, default=0)
    icon: Mapped[str] = mapped_column(String(50), default="atom")
    is_published: Mapped[bool] = mapped_column(Boolean, default=True)
    lessons: Mapped[list["Lesson"]] = relationship(back_populates="module", cascade="all, delete-orphan", order_by="Lesson.order_index")

class Lesson(Base):
    __tablename__ = "lessons"
    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    module_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("learning_modules.id"), nullable=False)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    content: Mapped[dict] = mapped_column(JSON, default=dict)
    order_index: Mapped[int] = mapped_column(Integer, default=0)
    xp_reward: Mapped[int] = mapped_column(Integer, default=50)
    estimated_minutes: Mapped[int] = mapped_column(Integer, default=10)
    module: Mapped["LearningModule"] = relationship(back_populates="lessons")
    progress: Mapped[list["LessonProgress"]] = relationship(back_populates="lesson", cascade="all, delete-orphan")

class LessonProgress(Base):
    __tablename__ = "lesson_progress"
    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("users.id"), nullable=False)
    lesson_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("lessons.id"), nullable=False)
    status: Mapped[ProgressStatus] = mapped_column(Enum(ProgressStatus), default=ProgressStatus.not_started)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    time_spent_seconds: Mapped[int] = mapped_column(Integer, default=0)
    lesson: Mapped["Lesson"] = relationship(back_populates="progress")

File: QuantumVerse_Organized/01_Frontend/src/types/learning.ts
export interface LearningModule {
  id: string;
  title: string;
  description: string;
  level: number;
  icon: string;
  lesson_count: number;
}

export interface Lesson {
  id: string;
  title: string;
  order_index: number;
  xp_reward: number;
  estimated_minutes: number;
  youtube_url?: string | null;
  content?: LessonContent;
}

export interface LessonContent {
  blocks: ContentBlock[];
}

export type ContentBlock =
  | { type: "text"; content: string }
  | { type: "heading"; level: 1 | 2 | 3; content: string }
  | { type: "math"; content: string }
  | { type: "code"; language: string; content: string }
  | { type: "callout"; variant: "info" | "warning" | "tip"; content: string }
  | { type: "circuit"; circuit_data: object };

export interface ProgressData {
  completed: string[];
  in_progress: string[];
}
